### PR TITLE
[improve][client]When namespace is deleted, the client is constantly reconnecting

### PR DESCRIPTION
**Pull request by poorbarcode**
### Motivation

If the namespace is deleted, the `cmd-lookup` will still keep retrying, because the error code sent to the client is `MetadataError`, and the client will retry when it gets `MetadataError`. No matter how many times it tries again, it will not succeed.

https://github.com/apache/pulsar/blob/6cc30c93a05bb16b7ef62ffebc15ceb5e6efb502/pulsar-broker/src/main/java/org/apache/pulsar/broker/lookup/TopicLookupBase.java#L250-L257

### Modifications

If the namespace is deleted
- the `cmd-lookup` will get the error code `TopicNotFound `
- the rest request will get the response code `404` 
- improve response message for `API topics` and `API namespaces`
- `TODO` I will write a test later

### Documentation

- [ ] `doc` 
- [ ] `doc-required` 
- [x] `doc-not-needed` 
- [ ] `doc-complete` 

### Matching PR in forked repository

PR in forked repository: 

- https://github.com/poorbarcode/pulsar/pull/30

### CODE REVIEW

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/v1/NonPersistentTopics.java [hunk 1]
```diff
@@ -121,7 +121,9 @@ public void getInternalStats(
     @Path("/{property}/{cluster}/{namespace}/{topic}/partitions")
     @ApiOperation(hidden = true, value = "Create a partitioned topic.",
             notes = "It needs to be called before creating a producer on a partitioned topic.")
-    @ApiResponses(value = {@ApiResponse(code = 403, message = "Don't have admin permission"),
+    @ApiResponses(value = {
+            @ApiResponse(code = 403, message = "Don't have admin permission"),
+            @ApiResponse(code = 404, message = "Namespace doesn't exist"),
```
Comment (michaeljmarshall) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/v1/NonPersistentTopics.java:126]: Nit:
```suggestion
            @ApiResponse(code = 404, message = "Namespace or topic does not exist"),
```

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/v1/Namespaces.java [hunk 2]
```diff
@@ -881,6 +881,7 @@ public void unloadNamespace(@Suspended final AsyncResponse asyncResponse, @PathP
     @ApiOperation(hidden = true, value = "Unload a namespace bundle")
     @ApiResponses(value = {
             @ApiResponse(code = 307, message = "Current broker doesn't serve the namespace"),
+            @ApiResponse(code = 404, message = "Namespace doesn't exist"),
```
Comment (michaeljmarshall) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/v1/Namespaces.java:884]: Nit: I think we probably want all of these to be like this:
```suggestion
            @ApiResponse(code = 404, message = "Namespace does not exist"),
```

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/v2/NonPersistentTopics.java [hunk 3]
```diff
@@ -541,7 +541,7 @@ public void getEntryFilters(@Suspended AsyncResponse asyncResponse,
     @Path("/{tenant}/{namespace}/{topic}/entryFilters")
     @ApiOperation(value = "Set entry filters for specified topic")
     @ApiResponses(value = {@ApiResponse(code = 403, message = "Don't have admin permission"),
-            @ApiResponse(code = 404, message = "Topic does not exist"),
+            @ApiResponse(code = 404, message = "Tenant or namespace or topic doesn't exist"),
```
Comment (michaeljmarshall) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/v2/NonPersistentTopics.java:544]: Is there a reason this string has `Tenant`, but most of the others do not?

**Review by michaeljmarshall — COMMENTED** [review]
The technical change looks good to me. I think we might want to clean up the annotations before merging, though.

**Review by Technoboy- — APPROVED** [review]
LGTM, please fix @michaeljmarshall suggestion.

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/v2/NonPersistentTopics.java (hunk 3 above)
Comment (poorbarcode) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/v2/NonPersistentTopics.java]: Already fixed. Could you take a look again?

**Review by nodece — APPROVED** [review]

**Comment by poorbarcode** [comment]
rebase master

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/web/PulsarWebResource.java [hunk 4]
```diff
@@ -875,7 +879,7 @@ public static CompletableFuture<ClusterDataImpl> checkLocalOrGetPeerReplicationC
                 if (!allowDeletedNamespace && policies.deleted) {
                     String msg = String.format("Namespace %s is deleted", namespace.toString());
                     log.warn(msg);
-                    validationFuture.completeExceptionally(new RestException(Status.PRECONDITION_FAILED,
+                    validationFuture.completeExceptionally(new RestException(Status.NOT_FOUND,
```
Comment (codelipenghui) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/web/PulsarWebResource.java:882]: It's incorrect here.
If `allowDeletedNamespace` is false. We should return `Status.PRECONDITION_FAILED`. No matter if `policies.deleted` is true or false.

Comment (poorbarcode) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/web/PulsarWebResource.java:882]: If we return `Status.PRECONDITION_FAILED` here, the error code will same as the scenario `if (policies.replication_clusters.isEmpty()`, then we can not determine whether the error was caused by `namespace deleted` when calling this method

Comment (codelipenghui) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/web/PulsarWebResource.java:882]: Oh, got it. I misunderstand the `allowDeletedNamespace`. I thought it was `forceDeleteNamespaceAllowed=false` in broker.conf.

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/web/PulsarWebResource.java [hunk 5]
```diff
@@ -822,8 +822,12 @@ protected void validateGlobalNamespaceOwnership(NamespaceName namespace) {
...
             throw e;
         } catch (Exception e) {
-            if (e.getCause() instanceof WebApplicationException) {
-                throw (WebApplicationException) e.getCause();
+            Throwable throwable = FutureUtil.unwrapCompletionException(e);
+            if (throwable instanceof WebApplicationException webApplicationException) {
+                throw webApplicationException;
+            }
+            if (throwable instanceof RestException restException){
+                throw restException;
```
[the remark this answers is not in the document]
Comment (poorbarcode) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/web/PulsarWebResource.java]: Already fixed

**Review by codelipenghui — APPROVED** [review]

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/v2/NonPersistentTopics.java (hunk 3 above)
Comment (michaeljmarshall) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/v2/NonPersistentTopics.java:544]: I don't think this was fixed. Why do we have `tenant` in this annotation but not in the annotation on line 192 above?

**Pull request merged**